### ISSUE TITLE
How to remove explicit public key algorithm info from Generated CSR from tpm2tss keys

### ISSUE THREAD

**Issue by samiransaha** [tpm2-software/tpm2-tss-engine#124]
Hi,

I am trying to generate CSR using tpm2 keys. 
I have found following commands to do that.

```
tpm2tss-genkey -a ecdsa -c nist_p256 tpm.key -v
openssl req -new -engine tpm2tss -keyform engine -out mycsr.csr -key tpm.key -subj "/CN=localhost"
```
But I am seeing that in public key information section in the generated CSR, all the nist_p256 parameters are explicitly mentioned. 
So, how can can tell tpm2tss to include public key algorithm named-identifier instead of explicit parameters?

By running `openssl req -in mycsr.csr -noout -text`
I am seeing like this(Partial Output):
```
            Public Key Algorithm: id-ecPublicKey
                Public-Key: (256 bit)
                pub:
                    04:98:48:97:9c:d0:eb:d5:90:42:18:31:d6:49:a2:
                    48:bd:1b:46:5e:ef:4b:f0:e1:81:61:5b:ee:c5:0c:
                    a2:89:0a:a7:12:1b:af:6c:87:b7:07:0c:4d:e3:db:
                    78:dd:02:e2:c6:99:4c:17:a4:b8:55:b2:81:1a:74:
                    33:25:9a:6f:53
                Field Type: prime-field
                Prime:
                    00:ff:ff:ff:ff:00:00:00:01:00:00:00:00:00:00:
                    00:00:00:00:00:00:ff:ff:ff:ff:ff:ff:ff:ff:ff:
                    ff:ff:ff
                A:
                    00:ff:ff:ff:ff:00:00:00:01:00:00:00:00:00:00:
                    00:00:00:00:00:00:ff:ff:ff:ff:ff:ff:ff:ff:ff:
                    ff:ff:fc
                B:
                    5a:c6:35:d8:aa:3a:93:e7:b3:eb:bd:55:76:98:86:
                    bc:65:1d:06:b0:cc:53:b0:f6:3b:ce:3c:3e:27:d2:
                    60:4b
                Generator (uncompressed):
                    04:6b:17:d1:f2:e1:2c:42:47:f8:bc:e6:e5:63:a4:
                    40:f2:77:03:7d:81:2d:eb:33:a0:f4:a1:39:45:d8:
                    98:c2:96:4f:e3:42:e2:fe:1a:7f:9b:8e:e7:eb:4a:
                    7c:0f:9e:16:2b:ce:33:57:6b:31:5e:ce:cb:b6:40:
                    68:37:bf:51:f5
                Order:
                    00:ff:ff:ff:ff:00:00:00:00:ff:ff:ff:ff:ff:ff:
                    ff:ff:bc:e6:fa:ad:a7:17:9e:84:f3:b9:ca:c2:fc:
                    63:25:51
                Cofactor:  1 (0x1)
                Seed:
                    c4:9d:36:08:86:e7:04:93:6a:66:78:e1:13:9d:26:
                    b7:81:9f:7e:90

```
I want like this:

```
            Public Key Algorithm: id-ecPublicKey
                Public-Key: (256 bit)
                pub:
                    04:98:48:97:9c:d0:eb:d5:90:42:18:31:d6:49:a2:
                    48:bd:1b:46:5e:ef:4b:f0:e1:81:61:5b:ee:c5:0c:
                    a2:89:0a:a7:12:1b:af:6c:87:b7:07:0c:4d:e3:db:
                    78:dd:02:e2:c6:99:4c:17:a4:b8:55:b2:81:1a:74:
                    33:25:9a:6f:53
                ASN1 OID: prime256v1
                NIST CURVE: P-256
```

Thanks in advance!

**Comment by AndreasFuchsSIT** [comment]
I honestly have no idea.
We basically just take everything from the TPM and push it into an OpenSSL key.
see here: https://github.com/tpm2-software/tpm2-tss-engine/blob/master/src/tpm2-tss-engine-ecc.c#L247

Do you have any additional information about what we should be setting in order to achieve your expected result ? We are already setting `ecgroup = EC_GROUP_new_by_curve_name(nid);` which I though should give us the curve and oid.

**Comment by samiransaha** [comment]
@AndreasFuchsSIT Sorry, currently I don't have any such information.

Can we use ecparam command in openssl to create key inside TPM, like below?
openssl ecparam -name secp256r1 -genkey -noout -out tpm.key -engine tpm2tss

ecparam can control the curve parameters format.

**Comment by AndreasFuchsSIT** [comment]
That's not the issue. We actually do create the keys via the curveid
https://github.com/tpm2-software/tpm2-tss-engine/blob/0bd2b64d4028b0db45ef4901e9f5af5f0c0a6770/src/tpm2-tss-engine-ecc.c#L456
and we also set the ecgroup for the key one we read it back from the TPM
https://github.com/tpm2-software/tpm2-tss-engine/blob/0bd2b64d4028b0db45ef4901e9f5af5f0c0a6770/src/tpm2-tss-engine-ecc.c#L271

So I'm currently not quite sure how to solve this in this side.
I guess the best would be to implement some small program and inspect the values of a key generated by openssl directly and one from the TPM. That's going to take some time though.

**Comment by samiransaha** [comment]
I have found this code in OpenSSL which manipulates the publickey info through ecparam command.
There is OPENSSL_EC_NAMED_CURVE flag we need to set in method while populating ecc key.

https://github.com/openssl/openssl/blob/master/crypto/ec/ec_pmeth.c:369
```
 else if (strcmp(type, "ec_param_enc") == 0) {
        int param_enc;
        if (strcmp(value, "explicit") == 0)
            param_enc = 0;
        else if (strcmp(value, "named_curve") == 0)
            param_enc = OPENSSL_EC_NAMED_CURVE;
        else
            return -2;
        return EVP_PKEY_CTX_set_ec_param_enc(ctx, param_enc);
```

**Comment by samiransaha** [comment]
I have able to make it working.
Following is the one liner change, if we want to make it default behaviour which I think should as there are lots of other crypto libraries which actually accepts only the named format.

```
diff --git a/src/tpm2-tss-engine-ecc.c b/src/tpm2-tss-engine-ecc.c
index 1d5688c..f2b03ce 100644
--- a/src/tpm2-tss-engine-ecc.c
+++ b/src/tpm2-tss-engine-ecc.c
@@ -278,6 +278,7 @@ populate_ecc(EC_KEY *key)
         EC_GROUP_free(ecgroup);
         return 0;
     }
+    EC_KEY_set_asn1_flag(key, OPENSSL_EC_NAMED_CURVE);
     EC_GROUP_free(ecgroup);

     x = BN_bin2bn(tpm2Data->pub.publicArea.unique.ecc.x.buffer,
```

**Comment by samiransaha** [comment]
@AndreasFuchsSIT Kindly help in merging the pull request https://github.com/tpm2-software/tpm2-tss-engine/pull/128.

**Comment by AndreasFuchsSIT** [comment]
Thanks a lot for the contribution !
Fixed via #128
